### ISSUE TITLE
Add support for resigning binaries in SignTool 

### ISSUE THREAD

**Issue by weshaggard** [dotnet/arcade#1204]
There are a number of times where we need to re-sign a strong named binary. For example for libraries where we have the full public/private key (i.e. open key) that we need to rewrite them after compile which causes any full signing the compiler did to become invalid. CoreFx has a number of cases like the GenFacades tool that has this issue. Another example would be adding optimization data to the binary after the fact and we need to resign the binary.

For keys like the Open key we cannot have the signing service strong name sign the binary because it is not a protected key with a protected certificate so the normal signing service will not sign it for us. We need to use either sn directly like we did in BuildTools (https://github.com/dotnet/buildtools/blob/master/src/Microsoft.DotNet.Build.Tasks/PackageFiles/sign.targets#L78) or add support for this in the SignTool itself. I'd much prefer the later as it keeps this signing logic centralized.

cc @JohnTortugo @tmat

**Comment by tmat** [comment]
I needed this for re-signing signed CoreFX binaries after merging IBC data into them.

This task clears the authenticode signature, which causes the sign tool to pass it to the signing service:
https://github.com/dotnet/roslyn-tools/blob/master/src/BuildTasks/Unsign.cs 

Does the signing service re-sign already signed binary if its signature doesn't match the content? If so then we should need Unsign task and instead pass a switch to the SignTool marking a file for resigning. 

In that case I think we should add a `ReSign` attribute that skips the "already signed" check in the SignTool for the particular file:

```
<FileSignInfo Include="..." ReSign="true"/>
```

**Comment by tmat** [comment]
Thinking about this more - I think the approach with `Unsign` task that we now use in Roslyn is actually simpler. Ideally, any tool that rewrites a binary should clear the signature. For tools that do not (like IBCMerge) we just call `Unsign` task on their output and then everything else proceeds as usual. There is no need for special signing configuration, the SignTool detects that the binary is unsigned and signs it as usual.

Re the Open key - that's an orthogonal issue. There we just need to either call `sn.exe` from the sign tool or implement strong-name signing in the SignTool. Then we just need to indicate in `StrongNameSignInfo` entry that tool should do the signing, versus let the signing service do it.

I think this could be done by using the path to the `.snk` file in `Include` instead of the certificate name. If the value ends with `.snk` the tool signs itself, otherwise it uses the service.

For example:

```
<StrongNameSignInfo Include="MsSharedLib72" PublicKeyToken="31bf3856ad364e35" CertificateName="Microsoft400" />
<StrongNameSignInfo Include="$(MSBuildThisFileDirectory)snks\Open.snk" PublicKeyToken="xxxx" CertificateName="Microsoft400" />
```

**Comment by markwilkie** [comment]
I like how @tmat is thinking about this too.

**Comment by weshaggard** [comment]
I agree we should include this support in SignTool task and avoid needing to create custom targets to workaround this.

**Comment by weshaggard** [comment]
For corefx scenarios we wouldn't need to worry about using an "unsign" task for our rewriters we would simply delaysign things with the open key like we do with all the other keys and then the SignTool would sign it. The only reason we had to hack around this in the past is because the signing service doesn't support this custom Open key so we needed to handle it directly.

**Comment by JohnTortugo** [comment]
I'm not quite familiar with the signing service so I don't know if all things considered this would be a good option or not: @weshaggard wouldn't things be simpler if the signing service were able to also sign the binaries with an Open Key? That way, seems to me, we wouldn't even have to change the SignTool.

**Comment by weshaggard** [comment]
You can certainly ask the owners of the signing service again but when we asked them in the past they wouldn't do it for any non-protected key. That is why I suggest we add support directly in our SignTool for it.

**Comment by markwilkie** [comment]
@weshaggard - Are you currently working around this today like you used to in build tools?

I'm wondering if these two issues should be moved to post Preview 2 as I'd like to stabilize. 

Thoughts?

**Comment by weshaggard** [comment]
No I'm not working around this as there isn't a great way to workaround it short of changing or stopping our use of SignTool. The issue is that we pass a set of nupkg packages to the SignTool and they it extracts all the binaries and signs them but for those binaries that use the Open key there is nothing in place in that pipeline that could strong name sign the binary. So to even workaround this we would need some sort of hook from the SignTool to do it. That is why I think we should just do the work in SignTool to fix it.

**Comment by markwilkie** [comment]
Yup, makes sense @weshaggard.  We'll get this work started.

**Comment by markwilkie** [comment]
cc/ @jcagme

**Comment by jcagme** [comment]
Is this currently breaking corefx? If so this should be fixed prior 12/4 (?)

**Comment by weshaggard** [comment]
It isn't breaking corefx exactly but it is currently breaking core-setup consuming those corefx bits as the signing validation is failing in the core-setup builds. So I do think we should fix this ASAP.

**Comment by jcagme** [comment]
Sure. @JohnTortugo is already on it.

**Comment by JohnTortugo** [comment]
Updating with the plan I've implemented.

The issue is:

“For keys like the Open key we cannot have the signing service strong name sign the binary because it is not a protected key with a protected certificate so the normal signing service will not sign it for us. We need to use either sn directly like we did in BuildTools (https://github.com/dotnet/buildtools/blob/master/src/Microsoft.DotNet.Build.Tasks/PackageFiles/sign.targets#L78) or add support for this in the SignTool itself. I'd much prefer the later as it keeps this signing logic centralized.”

The current plan is: 

Make Arcade SignTool be able to strong name sign files locally. For that we’ll leverage some configuration parameters that we already have (StrongNameSignInfo) and add the possibility of specifying the path to a .SNK file. If the path is specified all files with a matching Public Key Token will be locally signed (using sn.exe -R <path_to_file> <path_to_key>) otherwise the signing will flow to the remote signing service.

Example configuration:

```
<!— Current process: all files with PKT == “31bf3856ad364e35” will be strong-name signed with “MsSharedLib72” -->
<StrongNameSignInfo Include="MsSharedLib72" PublicKeyToken="31bf3856ad364e35" />

<!— What we are adding: all files with PKT == “1234567abcde” will be locally strong name signed with the “key.snk” Key -->
<StrongNameSignInfo Include=”/full/path/to/snk/key.snk" PublicKeyToken="1234567abcde" />

```
